### DISCUSSION TITLE
Fine-grained authorization

### DISCUSSION THREAD

**Discussion by mffap** [zitadel/zitadel#5623] [Ideas]
By default ZITADEL offers role-based access control (RBAC) and attribute-based access control (ABAC), as explained [in this discussion](https://github.com/zitadel/zitadel/discussions/5090). Managing access to specific resources and enforcing access based on policies is currently not in scope.

With this you have the option to enforce access to resources by defining roles for example in the form of functional user roles (eg, `admin`, `editor`, `reader`, ...) or related to an application feature (eg, `book:read`, `book:edit`, ...). These roles can be combined to enforce access to resources.

You can set custom metadata on users and organisations, as well as create custom claims with Actions, that can be used for ABAC. Even distributed claims would be possible where you request additional information in a third-party system and return the claims in the user information (eg, from an CRM system, training/certification tool, HR system, ...).

Managing application-specific resources (eg, books, shipping orders, IoT devices, documents) and the policies defining access to these resources based on certain attributes (eg, User-Sub, Roles, Claims, IP, Tenant/Organization, ...) could be integrated into ZITADEL, based on existing / familiar implementations of fine-grained authorization.

**Comment by mffap** [comment]
Currently the best way to implement fine-grained authorization is either by using custom application logic or, for larger scale projects, an available 3rd party tool like [warrant-dev/warrant](https://github.com/warrant-dev/warrant),[cerbos/cerbos](https://github.com/cerbos/cerbos), etc.

**Reply by kkajla12** [reply]
Thanks for the shoutout to Warrant @mffap! For folks interested in an open source solution for fine-grained authorization, I also want to point out that Warrant is open source. You can find our repository [here](https://github.com/warrant-dev/warrant).
